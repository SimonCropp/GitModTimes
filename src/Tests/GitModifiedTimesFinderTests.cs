using GitModTimes;
using VerifyXunit;
using Xunit;

[UsesVerify]
public class GitModifiedTimesFinderTests
{
    DateTimeOffset epoch = new(new(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc));

    public GitModifiedTimesFinderTests()
    {
        var testDir = Path.Combine(Path.GetTempPath(), "GitModTimes");
        Directory.CreateDirectory(testDir);
        foreach (var directory in Directory.EnumerateDirectories(testDir))
        {
            DirectoryHelper.DeleteDirectory(directory);
        }
    }

    [Fact]
    public Task Can_get_last_modified_dates()
    {
        var testDir = CreateTestDir("Can_get_last_modified_dates");
        using var repository = RepoBuilder.BuildTestRepository(testDir);
        var modifiedTimes = repository.GetTimes(testDir);
        return Verify(modifiedTimes);
    }

    [Fact]
    public Task Can_fix_dates()
    {
        var testDir = CreateTestDir("Can_fix_dates");
        using (var repository = RepoBuilder.BuildSimpleTestRepository(testDir))
        {
            repository.FixTimes(testDir, epoch);
        }
        return Verify(GetNonGitFiles(testDir));
    }

    static string CreateTestDir(string suffix)
    {
        return Path.Combine(Path.GetTempPath(), "GitModTimes", suffix);
    }

    [Fact]
    public Task Can_fix_dates_with_cutoff()
    {
        var testDir = CreateTestDir("Can_fix_dates_with_cutoff");
        using var repository = RepoBuilder.BuildSimpleTestRepository(testDir);
        var commit = repository.Commits
            .Skip(2)
            .First();
        repository.FixTimes(testDir, epoch, null, commit.Author.When);
        return Verify(GetNonGitFiles(testDir));
    }

    static IEnumerable<Tuple<string, DateTime>> GetNonGitFiles(string directory)
    {
        return from file in Directory.EnumerateFiles(directory, "*.*", SearchOption.AllDirectories)
            where !file.Contains(".git")
            select new Tuple<string, DateTime>(file, File.GetLastWriteTimeUtc(file));
    }

    [Fact]
    public Task Can_get_last_modified_dates_with_cutoff()
    {
        var testDir = CreateTestDir("Can_get_last_modified_dates_with_cutoff");
        using var repository = RepoBuilder.BuildSimpleTestRepository(testDir);
        var commit = repository.Commits
            .Skip(2)
            .First();
        var modifiedTimes = repository.GetTimes(testDir, null, commit.Author.When);

        return Verify(modifiedTimes);
    }
}