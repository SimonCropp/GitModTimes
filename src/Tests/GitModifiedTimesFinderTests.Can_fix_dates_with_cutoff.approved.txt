[
  {
    "Item1": "C:\One.md",
    "Item2": "1970-01-01T00:00:00Z"
  },
  {
    "Item1": "C:\Three.md",
    "Item2": "2012-04-17T06:22:35Z"
  },
  {
    "Item1": "C:\Two.md",
    "Item2": "2012-04-17T06:14:35Z"
  }
]